vyper/semantics/types/module.py

```diff
@@ -107,7 +107,8 @@ def _ctor_modifiability_for_call(self, node: vy_ast.Call, modifiability: Modifia
     def validate_implements(
         self, node: vy_ast.ImplementsDecl, functions: dict[ContractFunctionT, vy_ast.VyperNode]
     ) -> None:
-        fns_by_name = {fn_t.name: fn_t for fn_t in functions.keys()}
+        # only external functions can implement interfaces
+        fns_by_name = {fn_t.name: fn_t for fn_t in functions.keys() if fn_t.is_external}
```

Comment:
we should assert that all the functions are external, but leave it to the caller to filter

Reply:
addressed in: https://github.com/vyperlang/vyper/pull/4178/commits/e4f6622cd729eb7cdd26e6c8e8c455082f4500ad